hypothesis-python/src/hypothesis/internal/conjecture/utils.py

```diff
@@ -442,3 +443,40 @@ def reject(self):
                 self.data.mark_invalid()
             else:
                 self.force_stop = True
+
+
+@lru_cache()
+def _calc_p_continue(desired_avg, max_size):
+    """Return the p_continue which will generate the desired average size."""
+    if desired_avg == max_size:
+        return 1.0
+    p_continue = 1 - 1.0 / (1 + desired_avg)
+    if p_continue == 0 or max_size == float("inf"):
+        assert 0 <= p_continue < 1, p_continue
+        return p_continue
+    # For small max_size, the infinite-series p_continue is a poor approximation,
+    # and while we can't solve the polynomial a few rounds of iteration quickly
+    # gets us a good approximate solution in almost all cases (sometimes exact!).
+    while _p_continue_to_avg(p_continue, max_size) > desired_avg:
+        # This is impossible over the reals, but *can* happen with floats.
+        p_continue -= 0.0001
```

Comment:
In this case I think we can just return p_continue (or do nothing, thus skipping the zeroth iteration of the while loop)? If `1 - 1.0 / (1 + desired_avg)` is close enough to the correct answer that floating-point error causes p_continue to be an overestimate then we're close enough for practical purposes.

Reply:
I agree that the estimate is close enough that we could just use the initial p_continue.

However, it's cheap to make this adjustment, and that allows us to test that our binary search *always* maintains a strict upper bound.